client/extensions/woocommerce/products/products-variation-types-form/index.js

```diff
@@ -0,0 +1,157 @@
+			PropTypes.oneOfType( [ PropTypes.string, PropTypes.node ] ),
+			PropTypes.arrayOf( PropTypes.oneOfType( [ PropTypes.string, PropTypes.node ] ) )
+		] ),
+		product: PropTypes.shape( {
+			id: PropTypes.number.isRequired,
+			name: PropTypes.string.isRequired,
+			type: PropTypes.string.isRequired,
+		} ),
+		variations: PropTypes.arrayOf( PropTypes.shape( {
+			type: PropTypes.string.isRequired,
+			values: PropTypes.arrayOf( PropTypes.string )
+		} ) ),
+	};
+
+	static defaultProps = {
+		label: i18n.translate( 'Variation types' ),
+		description: i18n.translate(
+			'Let\'s add some variations! A common {{em}}variation type{{/em}} is color. The {{em}}values{{/em}} would be the colors the product is available in.',
+			{ components: { em: <em /> } }
+		),
+	};
+
+	constructor( props ) {
+		super( props );
+
+		this.state = {
+			variations: this.props.variations ? this.props.variations : this.getInitialFields(),
+			isVariation: props.product && 'variable' === props.product.type ? true : false,
+		};
+
+		this.addVariation = this.addVariation.bind( this );
+		this.renderInputs = this.renderInputs.bind( this );
+		this.handleToggle = this.handleToggle.bind( this );
+	}
+
+	getInitialFields() {
+		return [ this.getNewFields() ];
+	}
+
+	getNewFields() {
+		return {
+			type: '',
+			values: [],
+		};
+	}
+
+	handleToggle() {
+		this.setState( ( prevState ) => ( {
+			isVariation: ! prevState.isVariation,
+		} ) );
+	}
+
+	updateType( index, event ) {
+		event.preventDefault();
+		const newValue = event.target.value,
+			updatedVariations = clone( this.state.variations );
+		updatedVariations[ index ] = clone( updatedVariations[ index ] );
+		updatedVariations[ index ].type = newValue;
+		this.setState( { variations: updatedVariations } );
+	}
+
+	updateValues( index, value ) {
+		const updatedVariations = clone( this.state.variations );
+		updatedVariations[ index ] = clone( updatedVariations[ index ] );
+		updatedVariations[ index ].values = value;
+		this.setState( { variations: updatedVariations } );
+	}
+
+	renderInputs( variation, index ) {
+		return (
+			<div key={index} className="products-variation-types-form__fieldset">
+				<FormTextInput
+					placeholder={ i18n.translate( 'Color' ) }
+					value={ variation.type }
+					name="type"
+					onChange={ this.updateType.bind( this, index ) }
+					className="products-variation-types-form__field"
+				/>
+				<TokenField
+					value={ variation.values }
+					name="values"
+					onChange={ this.updateValues.bind( this, index ) }
+				/>
+			</div>
+		);
+	}
+
+	addVariation( event ) {
+		event.preventDefault();
+		const updatedVariations = this.state.variations.concat( [ this.getNewFields() ] );
+		this.setState( { variations: updatedVariations } );
+	}
+
+	render() {
+		const fields = this.state.variations,
+			inputs = fields.map( this.renderInputs ),
+			isNewProduct = this.props.product ? false : true;
+
+		let variationsForm = null;
+		if ( this.state.isVariation ) {
```

Comment:
In general, I'm not a fan of big if blocks in the render function. Mostly because it ambiguates the render in a big way.